Microsoft.Quantum.Canon/Utils/Multiplexor.qs

```diff
@@ -0,0 +1,285 @@
+            if(Length(qubits) == 0){
+                fail $"operation DiagonalUnitary -- Number of qubits must be greater than 0.";
+            }
+
+            // pad coefficients length to a power of 2.
+            let maxCoefficients = 2^(Length(qubits));
+            let padZeros = new Double[maxCoefficients - Length(coefficients)];
+            
+            // Compute new coefficients.
+            let (coefficients0, coefficients1) = MultiplexorZComputeCoefficients_(coefficients + padZeros);
+            MultiplexorZ(coefficients1, BigEndian(qubits[1..Length(qubits)-1]), qubits[0]);            
+            if(maxCoefficients == 2){
+                // Termination case
+                Exp([PauliI], 1.0 * coefficients0[0], qubits);
+            }
+            else{
+                DiagonalUnitary(coefficients0, BigEndian(qubits[1..Length(qubits)-1]));
+            }
+        }
+        adjoint auto
+        controlled auto
+        adjoint controlled auto
+    }
+
+    /// # Summary
+    /// Implementation step of multiply-controlled Z rotations.
+    /// # See Also
+    /// - Microsoft.Quantum.Canon.MultiplexorZ
+    function MultiplexorZComputeCoefficients_(coefficients: Double[]) : (Double[], Double[]) {
+        let newCoefficientsLength = Length(coefficients)/2;
+        mutable coefficients0 = new Double[newCoefficientsLength];
+        mutable coefficients1 = new Double[newCoefficientsLength];
+        for(idxCoeff in 0..newCoefficientsLength-1){
+            set coefficients0[idxCoeff] = 0.5 * (coefficients[idxCoeff] + coefficients[idxCoeff + newCoefficientsLength]);
+            set coefficients1[idxCoeff] = 0.5 * (coefficients[idxCoeff] - coefficients[idxCoeff + newCoefficientsLength]);
+        }
+        return (coefficients0, coefficients1);
+    }
+
+    
+    /// # Summary
+    /// Multiply-controlled unitary operation $U$ that applies a unitary $V_j$ 
+    /// when controlled by $n$-qubit number state $\ket{j}$.
+    ///
+    /// $U = \sum^{2^n-1}_{j=0}\ket{j}\bra{j}\otimes V_j$.
+    ///
+    /// # Input
+    /// ## unitaries
+    /// Array of up to $2^n$ unitary operations. The $j$th operation 
+    /// is indexed by the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// ## index
+    /// $n$-qubit control register that encodes number states $\ket{j}$ in
+    /// big-endian format.
+    ///
+    /// ## target
+    /// Generic qubit register that $V_j$ acts on.
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with identity elements if 
+    /// fewer than $2^n$ are specified. This implementation uses
+    /// $n-1$ ancilla qubits.
+    ///
+    /// # References
+    /// - Toward the first quantum simulation with quantum speedup
+    ///   Andrew M. Childs, Dmitri Maslov, Yunseong Nam, Neil J. Ross, Yuan Su
+    ///   https://arxiv.org/abs/1711.10980
+    operation MultiplexorUnitary<'T>(unitaries : ('T => () : Adjoint, Controlled)[], index: BigEndian, target: 'T) : () {
```

Comment:
Similar here, naming this as a verb would make it a bit more clear, I think. Perhaps `MultiplexOperations`?

Reply:
Renamed to `MultiplexOperations`